### DISCUSSION TITLE
Environment Working Group

### DISCUSSION THREAD

**Discussion by bundfussr** [pharmaverse/admiral#2556] [General]
Book Club Advanced R

Week 1: [https://adv-r.hadley.nz/functions.html](https://www.google.com/url?q=https://adv-r.hadley.nz/functions.html&sa=D&source=calendar&ust=1731401127792409&usg=AOvVaw3PodKgpuwnkFrle4xti2UV) (Stefan)
Week 2: [https://adv-r.hadley.nz/environments.html](https://www.google.com/url?q=https://adv-r.hadley.nz/environments.html&sa=D&source=calendar&ust=1731401127792409&usg=AOvVaw2WIg5DQOWK6MTU6JkE2rrG) (Daniel)
Week 3: [https://adv-r.hadley.nz/metaprogramming.html](https://www.google.com/url?q=https://adv-r.hadley.nz/metaprogramming.html&sa=D&source=calendar&ust=1731401127792409&usg=AOvVaw2AhoNXkmjK6f8XQfA1ayo7) & [https://adv-r.hadley.nz/meta-big-picture.html](https://www.google.com/url?q=https://adv-r.hadley.nz/meta-big-picture.html&sa=D&source=calendar&ust=1731401127792409&usg=AOvVaw36GEaatR7gkBprD9zoThEP) (Edoardo)
Week 4: [https://adv-r.hadley.nz/expressions.html](https://www.google.com/url?q=https://adv-r.hadley.nz/expressions.html&sa=D&source=calendar&ust=1731401127792409&usg=AOvVaw369lOTbLoeujEjc5rtspcg) (Ben)
Week 5: [https://adv-r.hadley.nz/quasiquotation.html](https://www.google.com/url?q=https://adv-r.hadley.nz/quasiquotation.html&sa=D&source=calendar&ust=1731401127792409&usg=AOvVaw2Lt9EeqESXtGO5YDMtuex4)
Week 6: [https://adv-r.hadley.nz/evaluation.html](https://www.google.com/url?q=https://adv-r.hadley.nz/evaluation.html&sa=D&source=calendar&ust=1731401127792409&usg=AOvVaw2RGBIEUJuzFrD1IwGSdyWj)

Week 7: Dive in to applications in {admiral}

**Comment by bundfussr** [comment]
### Week 1: Functions
Lexical scoping may cause issues when users try to "functionalize" their code. Consider for example:
```
library(tibble)
library(dplyr)
library(lubridate)
time_window <- 30

adsl <- tribble(
  ~USUBJID, ~TRTEDT,
  "1",      ymd("2024-10-23")
)

adsl %>%
  mutate(AEREPEDT = TRTEDT + time_window)
```
Users may put the script and derivations in functions:
```
create_adsl <- function() {
  library(tibble)
  library(dplyr)
  library(lubridate)
  time_window <- 30

  adsl <- tribble(
    ~USUBJID, ~TRTEDT,
    "1",      ymd("2024-10-23")
  )

  derive_aerepedt(adsl)
}

derive_aerepedt <- function(dataset) {
  dataset %>%
    mutate(AEREPEDT = TRTEDT + time_window)
}
```
But this fails due to lexical scoping:
```
> create_adsl()
Error in `mutate()`:
ℹ In argument: `AEREPEDT = TRTEDT + time_window`.
Caused by error:
! object 'time_window' not found
```
#### Possible Actions:
- Provide guidance for users how to functionalize their code. This could be a vignette in admiral and/or a chapter in [SAS and R](https://bayer-group.github.io/sas2r/).

**Reply by bms63** [reply]
Thanks for making a discussion!!  

I'm hesitant to make this a vignette in admiral as this is a larger R programming topic, but what about just a pharmaverse blog post as we can show specific admiral code and keep the clutter to a minimum in admiral.  This was my original intention with pharmaverse blog to showcase/discuss thing like this.

All for getting it included in SAS and R, but we don't have control of it and it might get out of date.  unless we were just thinking this would be a general topic that doesn't include admiral

**Reply by bundfussr** [reply]
I've renamed the heading to "Possible Actions".

I think we should discuss at the end (Week 7) what we implement and where.

**Reply by jimrothstein** [reply]
@bundfussr  
Good example of lexical scoping.
Will bring up during discussion:  
 **Is make functions self-contained**, ie add `time-window` as formal arugment to `derive_aerepedt()`, recommended fix??

**Comment by bms63** [comment]
So this is where I have always been lost with the environments here https://github.com/pharmaverse/admiral/blob/d5048cb514647586f4ceeaa0a7915c936151e00c/R/admiral_environment.R#L14

Why are we creating an empty environment?  Isn't  it already available as 7.2.3 says every environment has a Parent and the search path terminates with the empty environment

@bundfussr @ddsjoberg @manciniedoardo

**Reply by ddsjoberg** [reply]
> Why are we creating an empty environment? Isn't it already available as 7.2.3 says every environment has a Parent and the search path terminates with the empty environment

In this case, we're creating an environment with an empty _parent_ environment, and as you say, the search terminates at an empty environment. In our case, this is exactly what we want! Take a look at the search path below for the internal admiral env: the search does _not_ extend to the global env, which could potentially break our functions if the global env had objects matching the names of objects we were looking for.

``` r
rlang::env_parents(
  admiral:::admiral_environment, 
  last = rlang::empty_env()
)
#> [[1]] $ <env: empty>
```

<sup>Created on 2024-11-18 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

**Reply by bms63** [reply]
aaaahhhhhh...so if we did not have this then our functions would reach out into global environment looking for matching variables which would cause a breakdown of society and general anarchy.  got it!

so....next low-level question.  admiral functions do their admiralty thing and provide a result/dataframe and the admiral function places result/dataframe into the global environment so a user can do others thing to it. 

How do functions know to place this into the global environment for the user?

**Reply by ddsjoberg** [reply]
I would say that functions return objects from the _calling env_ (the environment from which the function is called), rather than to the global. (Although, functions are most often called from the global env.)

**Reply by bms63** [reply]
I guess I am struggling with environments!  Glad we have this group :) 

So a user runs an admiral function.  A admiral function has its own environment when called (as do all functions) and we restrict the search path with our package setup to not allow the admiral functions to search outside that enclosed environment.   [Side Question later 1]

The admiral function returns an object to "some environment".   If the admiral function is called from global environment then the final object will be returned to the global environment.  If called from another environment [Side Question later 2] then the object will be returned to that environment. 

Am I understanding correctly?

* [Side Question later 1] Is this a common practice in other packages to restrict search path of functions?
* [Side Question later 2]  What would be the other environments/situations?

**Reply by bundfussr** [reply]
> we restrict the search path with our package setup to not allow the admiral functions to search outside that enclosed environment.

I don't think that we restrict the search path. I.e., admiral is not different from other packages

**Reply by bms63** [reply]
So all packages have that pgkname_environment.R code in it?

**Reply by bundfussr** [reply]
> So all packages have that pgkname_environment.R code in it?

No. We use this special environment only for storing the admiral options. It does not affect the way how objects are searched for when an admiral function is executed.

**Reply by bms63** [reply]
So the admiral functions **_can search outside their enclosed environment_** if they can not find variables/information not already in that environment?

**Reply by bundfussr** [reply]
Yes, but this is not admiral specific. It is described in [6.4 Lexical scoping](https://adv-r.hadley.nz/functions.html#lexical-scoping).

**Reply by bms63** [reply]
Sorry I got confused with the ` admiral_environment <- new.env(parent = emptyenv())`  I thought this was restricting the search path of the admiral functions.   

Understood on lexical scoping that a function looks up through the environments if it can't find  variables/info needed in its enclosed environment and then terminates at the empty environment throwing an error.

**Reply by jimrothstein** [reply]
Probably tangential, but I think this (old -ish) R book has some diagrams and `realistic` examples for using environments.  I find drawing envirnonment hierarchy helpful.
-https://rstudio-education.github.io/hopr/environments.html

For understanding, namespace vs environments ... I've read and re-read this.   One day it will sink in.   More code examples might help.
- https://blog.thatbuthow.com/how-r-searches-and-finds-stuff/

**Reply by bms63** [reply]
Ohhh this post looks good.  Will try and take a look before the meeting

**Comment by bms63** [comment]
Do we create an R6 object as discussed in 7.6?  Would that help us with some of our environment issues?

**Comment by bms63** [comment]
https://github.com/pharmaverse/admiral/blob/269e021b49a49c7679738bf75a245df3cd6664fb/R/call_derivation.R#L128

` if (!identical(fixed_env, variable_env)) {`

This looks similar to section 7.3 Recursing over Environments - maybe a good discussion item as it is in context of admiral

**Comment by bms63** [comment]
For package loading - I've always "attached" packages in the beginning of the script just so it was clean, but reading 7.4.1 it feels like this is more important as if mid-script you "attach" other library the global environment changes

**Reply by bms63** [reply]
i guess namespaces help to solve this - but still the namespace won't have access to the package until its load mid-script

**Reply by bms63** [reply]
R avoids this problem by taking advantage of the function versus binding environment described above. Every function in a package is associated with a pair of environments: the package environment, which you learned about earlier, and the namespace environment.

The package environment is the external interface to the package. It’s how you, the R user, find a function in an attached package or with [::](https://rdrr.io/r/base/ns-dblcolon.html). Its parent is determined by search path, i.e. the order in which packages have been attached.

The namespace environment is the internal interface to the package. The package environment controls how we find the function; the namespace controls how the function finds its variables.

**Reply by jimrothstein** [reply]
I think `search()`  is order dependent because it just follows `chain` of attached environments., as you said above.

I don't think NS places a role, yet.

**Comment by bms63** [comment]
Just putting these here to discuss at next EWG

![image](https://github.com/user-attachments/assets/db8d6ae8-b562-4efd-b383-046a5d8a591c)

**Comment by jimrothstein** [comment]
Holiday Reading !??

Many years ago, one of my freshman college roommates walked in the door speaking high praises for **LISP.**  
For some of us, it takes a longer to catch on.  Over the weekend, I read through the
"EVAL" chapter (Chapter 3) of **Common LISP** ( https://www.cs.cmu.edu/~dst/LispBook/)
(Tourteszky, Introduction to Common LISP - 1990).  

"So that's what R is trying to do!"  

Seems that LISP is cleaner way to introduce the
many list-like structures of R and how they built, passed, evaluated
or otherwise manipulated.

Then I read a chapter of this book and it too greatly helped with R.   More light bulbs went on.

**Metaprogramming in R**
(https://www.amazon.com/exec/obidos/ASIN/1484228804/acmorg-20) by
Thomas Mailund (also:


More from Thomas Mailund:
https://mailund.github.io/r-programmer-blog/2018/09/20/scoping-rules-and-nse/)

**Comment by jimrothstein** [comment]
How R finds a dataset?

A month or so a ago we agreed to use the syntax:
```
x = package::dataset
```
Example, if one remembers the package (and spells it correctly)
```
pharamaversesdtm::ae
```

I don't quite understand the mechanism that serves up the dataset.   It is not exported object, I tried to find it in the NS of the package (had much trouble with this).   And we are using lazyload which has some affect on datasets, but I remain puzzled.

**Comment by jimrothstein** [comment]
This code came up today as an example of operator precedence.
(`!` is run last)
```
!x %in% y
```
I suggest another way to look at this code, equivalent.   Mentally (or with lobstr::ast) break up this code into a list of one function followed by its arguments (`!` takes only one argument)  

I think the list way of looking at R can simplify otherwise messy contructions. 

(https://github.com/jimrothstein/try_things_here/blob/main/NSE_project/900_not_function.md)

**Comment by bundfussr** [comment]
# Week 6: Evaluation

The main discussion point was why admiral functions expect expressions instead of quosures and how this affects the users. We considered the following example.

Function in admiral:
```
# This version expect expressions for the `cond` argument
derive_param_expr <- function(dataset, cond) {
  bind_rows(
    dataset,
    dataset %>% filter(!!enexpr(cond)) %>% mutate(PARAMCD = "New parameter")
  )
}

# This version expect quosures for the `cond` argument
derive_param_quo <- function(dataset, cond) {
  bind_rows(
    dataset,
    dataset %>% filter(!!enquo(cond)) %>% mutate(PARAMCD = "New parameter")
  )
}
```

A user creates a project specific function which calls the admiral function:
```
derive_param_project_expr <- function(dataset, param) {
  derive_param_expr(dataset, PARAMCD == param)
}

derive_param_project_quo <- function(dataset, param) {
  derive_param_quo(dataset, PARAMCD == param)
}
```

Now create some data to use the function:
```
data_good <- tibble(
  PARAMCD = c("A", "B", "C"),
  AVAL = c(10, 12, 42)
)
data_bad <- tibble(
  PARAMCD = c("A", "B", "C"),
  AVAL = c(10, 12, 42),
  param = c(1, 2, 3)
)
```

Using the "expression" version usually fails (`data_good`) or produce wrong results (`data_bad`):
```
> derive_param_project_expr(data_good, "B")
Error in `filter()`:
ℹ In argument: `PARAMCD == param`.
Caused by error:
! object 'param' not found
Run `rlang::last_trace()` to see where the error occurred.
> derive_param_project_expr(data_bad, "B")
# A tibble: 3 × 3
  PARAMCD  AVAL param
  <chr>   <dbl> <dbl>
1 A          10     1
2 B          12     2
3 C          42     3
```

Using the "quosure" version works in most cases (`good_data`) but produce wrong results in some cases (`data_bad`):
```
> derive_param_project_quo(data_good, "B")
# A tibble: 4 × 2
  PARAMCD        AVAL
  <chr>         <dbl>
1 A                10
2 B                12
3 C                42
4 New parameter    12
> derive_param_project_quo(data_bad, "B")
# A tibble: 3 × 3
  PARAMCD  AVAL param
  <chr>   <dbl> <dbl>
1 A          10     1
2 B          12     2
3 C          42     3
```

Both versions doesn't work correctly because `param` in `PARAMCD == param` is ambiguous. The difference is when the issue occurs. In the "expression" version it is most likely noticed when the project specific function is developed. In the "quosure" version it is most likely noticed by the user but not by the developer of the project specific function.

**Reply by jimrothstein** [reply]
Instead of !!, consider pronouns **.env** and **.data**
SEE
https://rlang.r-lib.org/reference/injection-operator.html

**Reply by bundfussr** [reply]
Using the pronouns `.env` and `.data` would work only if the admiral function expect quosures instead of expressions.

**Reply by jimrothstein** [reply]
This video is about RACKET, of lisp origin but R uses a few of the same words.
Here Prof Kristopher Micinski explains **expression**  must evaluate to a value and `if` is a  `form`  (in R, **`if`** is speical) .  Forms do not need to evaluate to a value (infinite loops, or may evaluate only 1 arguments.)

May not improve R skills, but gives me greater comfort with some of the ideas R uses.


[Kristopher Micinski](https://www.youtube.com/@krismicinski)
